src/libopenrave/kinbody.cpp

```diff
@@ -4312,50 +4304,120 @@ void KinBody::_ComputeInternalInformation()
 
     // compute the all-pairs shortest paths
     {
-        _vAllPairsShortestPaths.resize(_veclinks.size()*_veclinks.size());
-        FOREACH(it,_vAllPairsShortestPaths) {
+        // Preallocate to fit our NxN joint map
+        _vAllPairsShortestPaths.resize(_veclinks.size() * _veclinks.size());
+
+        // Default each entry to a pair of invalid joint indices
+        FOREACH(it, _vAllPairsShortestPaths) {
             it->first = -1;
             it->second = -1;
         }
-        vector<uint32_t> vcosts(_veclinks.size()*_veclinks.size(),0x3fffffff); // initialize to 2^30-1 since we'll be adding
-        for(size_t i = 0; i < _veclinks.size(); ++i) {
-            vcosts[i*_veclinks.size()+i] = 0;
+
+        // All of our arrays are essentially 2d look up tables, so create a wrapper to generate an array index from a 2d point
+        const size_t linksSize = _veclinks.size();
+        auto MakeIndex = [&](size_t x, size_t y) -> size_t { return x * linksSize + y; };
```

Comment:
Can you elaborate on the reason behind these rules? 

With regard to lambdas (which do not appear to be called out in cpp.md?) - this expression is heavily repeated, so extracting and naming it makes it clearer what's going on / reduces risk of programmer error. Of the other options considered:
- `#define`s aren't naturally scoped, and require an #undef
- A separate function semantically separates the logic from where it's used and can result in lots of trivial helper functions littering the namespace.

Lambdas are naturally scoped, can occur next to the logic they are part of, and are (for simple cases such as this) trivially flattened by the compiler.

Similarly curious behind the reason for avoiding `auto`, especially when I see a fair amount of `decltype`, which is almost identical in function but slower for the compiler to process.

Reply:
One reason to avoid `auto` that I find useful is to force the developer to think about the type. Sometimes they don't realize the resulting type after a long line of calls. They may not be paying enough attention to avoid copy for example. It also makes code review easier, since it shows the developer thought about it hard, and reviewers do not need to do the job of a compiler to figure out the type again in their heads.

Reply:
It's better to be explicit than implicit. Adding trivial functions and typing out the types gives clarity to the code. Clarity, which I think is paramount as a codebase expands. 
I want to avoid a philosophical argument here, so I'm not going to comment on this further, please follow the C++ coding guidelines that you can find on gitlab.
Thanks